packages/vats/src/transfer.js

```diff
@@ -0,0 +1,242 @@
+   * sending messages which may violate the transfer protocol directly to the
+   * `targetHost`.
+   *
+   * @param {Parameters<typeof makeTransferInterceptorKit>} args
+   */
+  const makeTransferInterceptor = (...args) =>
+    makeTransferInterceptorKit(...args).public;
+  return makeTransferInterceptor;
+};
+harden(prepareTransferInterceptor);
+
+const TransferMiddlewareI = M.interface('TransferMiddleware', {
+  registerTap: M.callWhen(
+    M.string(),
+    M.remotable('TransferInterceptor'),
+  ).returns(M.remotable('TargetRegistration')),
+  registerActiveTap: M.callWhen(
+    M.string(),
+    M.remotable('TransferInterceptor'),
+  ).returns(M.remotable('TargetRegistration')),
+  unregisterTap: M.callWhen(M.string()).returns(),
+});
+
+/**
+ * @callback RegisterTap
+ * @param {string} target String identifying the bridge target.
+ * @param {ERef<import('./bridge-target.js').TargetApp>} tap The "application
+ *   tap" to register for the target.
+ */
+
+/**
+ * @param {import('@agoric/base-zone').Zone} zone
+ * @param {ReturnType<typeof prepareTransferInterceptor>} makeTransferInterceptor
+ */
+const prepareTransferMiddleware = (zone, makeTransferInterceptor) =>
+  zone.exoClass(
+    'TransferMiddleware',
+    TransferMiddlewareI,
+    /**
+     * The TransferMiddleware attenuates a `targetHost` and `targetRegistry` by
+     * wrapping TargetApps with a TransferInterceptor.
+     *
+     * @param {object} arg0
+     * @param {ERef<import('./bridge-target.js').TargetHost>} arg0.targetHost
+     * @param {ERef<import('./bridge-target.js').TargetRegistry>} arg0.targetRegistry
+     */
+    ({ targetHost, targetRegistry }) => ({ targetHost, targetRegistry }),
+    {
+      /**
+       * Register a tap to intercept the vtransfer target account address
+       * messages, without granting that tap the ability to delay or interfere
+       * with the underlying IBC transfer protocol.ß
+       *
+       * @type {RegisterTap}
+       */
+      async registerTap(target, tap) {
+        const { targetHost, targetRegistry } = this.state;
+
+        // Wrap the tap so that its upcall results determine how to contact the
+        // targetHost.  Never allow the tap to send to the targetHost directly.
+        const interceptor = makeTransferInterceptor(targetHost, tap);
+        return E(targetRegistry).register(target, interceptor);
+      },
+      /**
+       * Similar to `registerTap`, but allows the tap to inject async
+       * ßacknowledgements in the IBC transfer protocol.
+       *
+       * @type {RegisterTap}
+       */
+      async registerActiveTap(target, tap) {
+        const { targetHost, targetRegistry } = this.state;
+
+        // Wrap the tap and allow it to modify async acknowledgements.  Never
+        // allow the tap to send to the targetHost directly.
+        const interceptor = makeTransferInterceptor(targetHost, tap, true);
+        return E(targetRegistry).register(target, interceptor);
```

Comment:
**WARNING!!!**
I think this direct response allows consumers to bypass the `makeTransferInterceptor` protections, e.g.
```js
const rawRegistration =
  // makes a new TransferInterceptor `interceptor` around `unwrappedTap` and
  // maps to E($transferMiddlewareState.targetRegistry).register(addr, interceptor)
  await E(transferMiddleware).registerTap(addr, unwrappedTap);

// maps to E($transferMiddlewareState.targetRegistry).reregister(addr, unwrappedTap)
E(rawRegistration).updateTargetApp(unwrappedTap);
```

If you really want the interface to be [intercepted] middleware, then I think the registry itself must be configured with a corresponding function such that the current relationship is inverted and transfer-proposal.js looks like e.g.
```js
// interceptorFactory has a `wrapApp` method with the same parameters as
// `makeTransferInterceptor` in transfer.js.
const interceptorFactory = E(vats.transfer).makeInterceptorFactory();

// The third argument is to be used by `register` and `reregister` for replacing `app`.
// Those need to grow a new parameter for passing options, e.g.
// async register(target, app, ...wrapOpts) { …;
//   app = await E(appWrapper).wrapApp(targetHost, app, ...wrapOpts));
// …; }
const registry = await E(vats.transfer).provideBridgeTargetKit(
  bridgeChannel,
  VTRANSFER_IBC_EVENT,
  interceptorFactory,
);

// `registerTap(addr, app)` calls `E(registry).register(addr, app)`.
// `registerActiveTap(addr, app)` calls `E(registry).register(addr, app, true)`,
// in which `register` passes `true` to the interceptorFactory `wrapApp` method
// (what is now `makeTransferInterceptor`) after `targetHost` and `app`.
const transferMiddleware = await E(vats.transfer).makeTransferMiddleware(registry);
```

Reply:
The refactoring is now up as a PR against the source branch of _this_ PR: https://github.com/schnetzlerjoe/agoric-sdk/pull/2